File: tests/pages/DemoPage.js
/**
 * Page Object Model の例
 * ページの操作をクラスにカプセル化
 */
class DemoPage {
  constructor(page) {
    this.page = page

    // セレクタを一箇所で管理
    this.counter = page.locator('text=/^\\d+$/').first()
    this.incrementButton = page.getByRole('button', { name: '+1' })
    this.resetButton = page.getByRole('button', { name: 'リセット' })
    this.redButton = page.getByRole('button', { name: '赤', exact: true })
    this.greenButton = page.getByRole('button', { name: '緑', exact: true })
    this.blueButton = page.getByRole('button', { name: '青', exact: true })
  }

  // ページ固有の操作をメソッドにする
  async goto() {
    await this.page.goto('/demo')
  }

  async incrementCounter(times = 1) {
    for (let i = 0; i < times; i++) {
      await this.incrementButton.click()
    }
  }

  async resetCounter() {
    await this.resetButton.click()
  }

  async selectColor(color) {
    const colorMap = {
      '赤': this.redButton,
      '緑': this.greenButton,
      '青': this.blueButton,
    }
    await colorMap[color].click()
  }

  async getCounterValue() {
    return await this.counter.textContent()
  }

  async expectCounterToBe(value) {
    await this.page.expect(this.counter).toHaveText(value.toString())
  }
}

module.exports = { DemoPage }

File: tests_py/pages/demo_page.py
"""
Page Object Model - デモページ
Python版のPage Object実装例
"""
from playwright.sync_api import Page, Locator


class DemoPage:
    """デモページのPage Objectクラス"""

    def __init__(self, page: Page):
        self.page = page

        # セレクタを定義
        self.counter: Locator = page.locator("text=/^\\d+$/").first
        self.increment_button: Locator = page.get_by_role("button", name="+1")
        self.reset_button: Locator = page.get_by_role("button", name="リセット")
        self.red_button: Locator = page.get_by_role("button", name="赤", exact=True)
        self.green_button: Locator = page.get_by_role("button", name="緑", exact=True)
        self.blue_button: Locator = page.get_by_role("button", name="青", exact=True)

    def goto(self) -> None:
        """デモページに移動"""
        self.page.goto("/demo")

    def increment_counter(self, times: int = 1) -> None:
        """カウンターを指定回数増やす"""
        for _ in range(times):
            self.increment_button.click()

    def reset_counter(self) -> None:
        """カウンターをリセット"""
        self.reset_button.click()

    def select_color(self, color: str) -> None:
        """
        色を選択

        Args:
            color: "赤", "緑", または "青"
        """
        color_map = {
            "赤": self.red_button,
            "緑": self.green_button,
            "青": self.blue_button,
        }
        if color not in color_map:
            raise ValueError(f"Invalid color: {color}. Must be one of: 赤, 緑, 青")

        color_map[color].click()

    def get_counter_value(self) -> str:
        """現在のカウンター値を取得"""
        return self.counter.text_content() or "0"

    @property
    def url(self) -> str:
        """現在のページURL"""
        return self.page.url
